client/src/main/java/org/eclipse/hono/client/impl/CommandClientImpl.java

```diff
@@ -0,0 +1,107 @@
+package org.eclipse.hono.client.impl;
+
+import java.net.HttpURLConnection;
+import java.util.Objects;
+import java.util.UUID;
+
+import org.eclipse.hono.client.CommandClient;
+import org.eclipse.hono.client.StatusCodeMapper;
+import org.eclipse.hono.config.ClientConfigProperties;
+import org.eclipse.hono.util.CacheDirective;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.BufferResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.proton.ProtonConnection;
+
+/**
+ * A Vertx-Proton based client for Hono's Command and Control API.
+ *
+ */
+public class CommandClientImpl extends AbstractRequestResponseClient<BufferResult> implements CommandClient {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CommandClientImpl.class);
+
+    CommandClientImpl(final Context context, final ClientConfigProperties config, final String tenantId, final String deviceId) {
+        super(context, config, tenantId, deviceId);
+    }
+
+    @Override
+    protected String getName() {
+        return CommandConstants.COMMAND_ENDPOINT;
+    }
+
+    @Override
+    protected String createMessageId() {
+        return String.format("cmd-client-%s", UUID.randomUUID());
+    }
+
+    @Override
+    protected BufferResult getResult(final int status, final Buffer payload, final CacheDirective cacheDirective) {
+        return BufferResult.from(status, payload);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Future<Buffer> sendCommand(final String command, final Buffer data) {
+
+        Objects.requireNonNull(command);
+        Objects.requireNonNull(data);
+
+        final Future<BufferResult> responseTracker = Future.future();
+        createAndSendRequest(command, data, responseTracker.completer());
+
+        return responseTracker.map(response -> {
+            switch (response.getStatus()) {
+            case HttpURLConnection.HTTP_OK:
```

Comment:
this does not match the interface which says that the future succeeds for all 2xx codes.